### PR TITLE
fix: self-heal manifest-unreferenced branch forks (stop wedged branches)

### PR DESCRIPTION
**Pull request by ragnorc**
## What & why

A first write to a table on a branch lazily forks it via Lance `create_branch` — a durable, two-phase op that advances Lance state **before** the atomic manifest publish — so a crash, deploy restart, or cancelled request future between the fork and the publish left a fully-formed branch ref the manifest never referenced, wedging every subsequent write to that table on that branch with `"orphaned table state … incomplete prior delete; run cleanup"` (and `cleanup` couldn't even fix it, since the branch was still live). This makes the per-table fork **derived state of the manifest**: the fork now returns a typed `ForkOutcome` and self-heals a manifest-unreferenced leftover (reclaim + re-fork under the write queue) on the next write across load/mutate/merge, with `cleanup`'s reconciler broadened to a per-table authority test as the guaranteed backstop.

## Backing issue / RFC

- [x] **Maintainer fix** for a reproduced durability defect (interrupted branch fork wedges writes). No public issue/RFC; per the template note, maintainer internal process applies. Deterministic local repro added as the lead test commit.

## Checklist

- [x] Change is focused (one logical change) — the fork-reclaim fix; a small related stale-comment cleanup rides in its own `chore:` commit
- [x] Tests added/updated for behavior changes — red→green regression (`writes.rs`), reconcile coverage (`maintenance.rs`), failpoints flip + preserved retryable test
- [x] Public docs updated if user-facing surface changed (or N/A) — N/A user-facing; dev docs (`writes.md`, `invariants.md`) updated
- [x] Reviewed against docs/dev/invariants.md — no Hard Invariant weakened; uses the authority-derived reconciler pattern the invariants prefer

## Notes for reviewers

Reclaim is safe because branch first-writes now acquire the per-`(table, branch)` write queues up front (held through publish), so no in-process writer can be mid-fork; cross-process in-flight forks remain the already-documented one-winner-CAS gap (noted in `invariants.md`). Commits are file-granular and ordered red→green (`3edbe2e` is the RED test, `6e3204d` the fix) since interactive hunk staging wasn't available. Only steady-state cost is added serialization of concurrent first-writes to the *same* `(table, branch)` — which already conflict at commit; main-branch and already-forked writes are unaffected.

---

> [!NOTE]
> **High Risk**
> Changes destructive fork reclaim, manifest authority checks, and write-queue ordering on load/mutate/merge/cleanup paths; mistakes could delete live forks or leave writes unserialized, though tests and failpoints target those cases.
> 
> **Overview**
> Fixes a durability bug where an interrupted first-write fork (Lance `create_branch` before manifest publish) left a branch ref the manifest never referenced, wedging later writes with “run cleanup” even on live branches.
> 
> **Write path:** `fork_branch_from_state` now returns `ForkOutcome::RefAlreadyExists` instead of a hard error; the engine runs `reclaim_orphaned_fork_and_refork` after `classify_fork_ref` confirms the ref is an orphan from fresh manifest authority. Load and mutate acquire per-`(table, branch)` write queues up front for all touched tables (including cascade-delete edges) and pass **held guards** into `commit_all` so reclaim cannot race in-process forks and cannot self-deadlock on re-acquire.
> 
> **Cleanup:** `reconcile_orphaned_branches` treats orphans **per table** (live branch but table not placed on it), re-validates under the write queue via the same classifier, and skips on `Indeterminate` reads.
> 
> Dev docs and tests are updated (self-heal regressions, failpoint flips); minor comment/README tidy.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1eea371e07ed0c1868a85eb68e2e3a29f56fbbb1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Greptile Summary</h3>

This PR fixes a durability wedge on the branch first-write path: Lance's `create_branch` advances storage state before the manifest publish, so a crash or cancelled future left a fully-formed branch ref the manifest never referenced, causing every subsequent write to that table on that branch to error with "incomplete prior delete; run cleanup." The fix makes the fork a **self-healing derived-state operation**: `fork_branch_from_state` now returns a typed `ForkOutcome` rather than surfacing a collision as an error, `reclaim_orphaned_fork_and_refork` force-deletes and re-forks under fresh manifest authority, and `reconcile_orphaned_branches` is broadened to a per-table authority test (origin-2 orphans on live branches).

- **Write path:** `mutation.rs` and `loader/mod.rs` pre-acquire per-`(table, branch)` write queues for all touched tables (including cascade edges for node deletes) before the first fork and hold them through the publish; `commit_all` promotes the guard-coverage check from `debug_assert` to an all-builds error.
- **Cleanup:** `reconcile_orphaned_branches` now tests each `(table, branch)` pair against fresh manifest authority under the write queue via the shared `classify_fork_ref` classifier (`Orphan` / `Legitimate` / `Indeterminate`), preventing the stale-`live_branches` shortcut from destroying a legitimately-published fork on a newly-created branch.
- **Tests:** Red→green regression in `writes.rs`, per-table cleanup coverage in `maintenance.rs`, and failpoint tests pinning the `Indeterminate` skip, reconcile converge, and snapshot-failure caching paths in `failpoints.rs`.

<h3>Confidence Score: 5/5</h3>

Safe to merge. The destructive fork reclaim requires two independent gates to agree (in-process write queue serialization + fresh manifest authority via classify_fork_ref returning Orphan) before any force-delete runs; the Indeterminate path consistently skips rather than destroys on ambiguity. All previously flagged issues have been addressed.

The core fix is correct by construction: ForkOutcome replaces an opaque error with a typed signal, classify_fork_ref consolidates the 'safe to destroy?' decision into one site shared by both reclaim paths, and the write-queue pre-acquisition closes the in-process race. Previous review findings (debug_assert promotion, string-based error matching, Err(_) => destructive action, stale live_branches shortcut) are all addressed. The remaining inline comment concerns a rare diagnostic edge case in the Legitimate error arm where expected and actual versions could coincide — retryable and safe, just slightly misleading. Test coverage spans red→green regression, failpoint-injected Indeterminate/skip/converge paths, and the new per-table cleanup authority test.

table_ops.rs (reclaim_orphaned_fork_and_refork) — the Legitimate error arm's second fresh_snapshot_for_branch call can produce an expected==actual version mismatch if the read fails between classify and the version lookup; suggested fix inline.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| crates/omnigraph/src/db/omnigraph/table_ops.rs | Core of the fix: adds ForkRefStatus classifier, classify_fork_ref (shared by both reclaim sites), and reclaim_orphaned_fork_and_refork; write path now calls db.fork_dataset_from_entry_state (self-healing wrapper) instead of the raw fork function |
| crates/omnigraph/src/db/omnigraph/optimize.rs | reconcile_orphaned_branches broadened to per-table authority: origin-2 orphans (manifest-unreferenced fork on a live branch) now detected via per-branch snapshot + re-validated under the write queue via classify_fork_ref before any force-delete |
| crates/omnigraph/src/exec/staging.rs | commit_all gains held_guards parameter for the fork path; the pre-held coverage check is promoted from debug_assert to an all-builds error, safely failing the write if the pre-computed touched set diverges from the actual staged/inline set |
| crates/omnigraph/src/exec/mutation.rs | mutate_as now lowers IR up-front (lower_named_mutation), computes touched_table_keys (including cascade edges for node deletes), pre-acquires per-(table, branch) write queues when any table needs forking, and threads the guards into commit_all |
| crates/omnigraph/src/loader/mod.rs | load_jsonl_reader gains the same up-front fork-queue acquisition pattern as the mutation path; when any touched table needs forking, guards for all touched tables are acquired before Phase 2 and threaded through commit_all |
| crates/omnigraph/src/storage_layer.rs | ForkOutcome<D> added; fork_branch_from_state return type changed to Result<ForkOutcome<SnapshotHandle>>; RefAlreadyExists is now a typed signal (disambiguated via list_branches) rather than an opaque error |
| crates/omnigraph/src/table_store.rs | fork_branch_from_state disambiguates a failed create_branch: confirms via list_branches that the ref actually exists before returning RefAlreadyExists, propagating the raw error if the ref is absent (zombie or transient failure) |
| crates/omnigraph/tests/writes.rs | Two new regression tests: first_write_self_heals_manifest_unreferenced_fork_on_live_branch (via both load and mutate) and branch_cascade_delete_forks_node_and_edges_under_held_queues covering up-front queue coverage for cascade edge tables |
| crates/omnigraph/tests/failpoints.rs | Tests flipped from expecting wedge to expecting self-heal; new tests cover Indeterminate authority path (classify.fresh_read failpoint), reconcile skip-on-indeterminate then converge, and the caching test for live-branch snapshot resolution failure |
| crates/omnigraph/tests/maintenance.rs | New cleanup_reconciles_live_branch_orphan_fork_but_keeps_legitimate_fork test verifies origin-2 per-table authority: orphaned Person fork reclaimed while legitimately-forked Company on same live branch is preserved |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Caller as mutate_as / load_as
    participant WQ as WriteQueue
    participant OB as open_owned_dataset_for_branch_write
    participant FS as fork_dataset_from_entry_state (Omnigraph wrapper)
    participant TS as TableStore.fork_branch_from_state
    participant CFR as classify_fork_ref
    participant ROR as reclaim_orphaned_fork_and_refork
    participant CA as commit_all

    Caller->>WQ: acquire_many(touched tables x branch) [if needs_fork]
    WQ-->>Caller: guards held

    Caller->>OB: open for branch write
    OB->>FS: fork (source to active_branch)
    FS->>TS: create_branch(target_branch, source_version)

    alt RefAlreadyExists
        TS-->>FS: ForkOutcome::RefAlreadyExists
        FS->>ROR: reclaim_orphaned_fork_and_refork
        ROR->>CFR: classify_fork_ref(fresh authority)
        CFR-->>ROR: Orphan / Legitimate / Indeterminate
        alt Orphan
            ROR->>TS: force_delete_branch + re-fork
            ROR-->>FS: Ok(SnapshotHandle)
        else Legitimate / Indeterminate
            ROR-->>Caller: retryable error
        end
    else Created
        TS-->>FS: ForkOutcome::Created(dataset)
        FS-->>OB: Ok(SnapshotHandle)
    end

    OB-->>Caller: (dataset, Some(active_branch))
    Caller->>CA: commit_all(held_guards)
    CA->>CA: coverage check (queue_keys subset acquired_keys)
    CA-->>Caller: (updates, versions)
    Note over Caller,WQ: guards released at end of scope
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `crates/omnigraph/src/exec/mutation.rs`, line 456-469 ([link](https://github.com/modernrelay/omnigraph/blob/37f5af587afecc9607b9bead6e9bce30805852ec/crates/omnigraph/src/exec/mutation.rs#L456-L469)) 

   <a href="#">P2</a> **`needs_fork = false` leaves a fork-without-queue gap on cross-process branch recreate**

   If `needs_fork` is `false` at snapshot time (all touched tables appear already-forked), no guards are pre-acquired. If a foreign-process then deletes and recreates the branch between this snapshot fetch and `execute_named_mutation`, some tables become unforked; `open_owned_dataset_for_branch_write` re-detects this and calls `db.fork_dataset_from_entry_state`, which may reach `reclaim_orphaned_fork_and_refork` — without the per-`(table, branch)` queue held. The invariant comment on `reclaim_orphaned_fork_and_refork` states callers "MUST already hold" the queue.

   This is acknowledged in `invariants.md` as the one-winner-CAS gap and leads to retries rather than corruption, so it's not a blocking concern. Worth calling out explicitly since the comment on `reclaim_orphaned_fork_and_refork` reads as an unconditional precondition but actually has this cross-process carve-out.

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20crates%2Fomnigraph%2Fsrc%2Fexec%2Fmutation.rs%0ALine%3A%20456-469%0A%0AComment%3A%0A**%60needs_fork%20%3D%20false%60%20leaves%20a%20fork-without-queue%20gap%20on%20cross-process%20branch%20recreate**%0A%0AIf%20%60needs_fork%60%20is%20%60false%60%20at%20snapshot%20time%20%28all%20touched%20tables%20appear%20already-forked%29%2C%20no%20guards%20are%20pre-acquired.%20If%20a%20foreign-process%20then%20deletes%20and%20recreates%20the%20branch%20between%20this%20snapshot%20fetch%20and%20%60execute_named_mutation%60%2C%20some%20tables%20become%20unforked%3B%20%60open_owned_dataset_for_branch_write%60%20re-detects%20this%20and%20calls%20%60db.fork_dataset_from_entry_state%60%2C%20which%20may%20reach%20%60reclaim_orphaned_fork_and_refork%60%20%E2%80%94%20without%20the%20per-%60%28table%2C%20branch%29%60%20queue%20held.%20The%20invariant%20comment%20on%20%60reclaim_orphaned_fork_and_refork%60%20states%20callers%20%22MUST%20already%20hold%22%20the%20queue.%0A%0AThis%20is%20acknowledged%20in%20%60invariants.md%60%20as%20the%20one-winner-CAS%20gap%20and%20leads%20to%20retries%20rather%20than%20corruption%2C%20so%20it's%20not%20a%20blocking%20concern.%20Worth%20calling%20out%20explicitly%20since%20the%20comment%20on%20%60reclaim_orphaned_fork_and_refork%60%20reads%20as%20an%20unconditional%20precondition%20but%20actually%20has%20this%20cross-process%20carve-out.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=modernrelay%2Fomnigraph&pr=231&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a>
</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%201%20code%20review%20issue.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%201%0Acrates%2Fomnigraph%2Fsrc%2Fdb%2Fomnigraph%2Ftable_ops.rs%3A726-738%0AThe%20%60Legitimate%60%20arm%20calls%20%60fresh_snapshot_for_branch%60%20a%20second%20time%20after%20%60classify_fork_ref%60%20already%20called%20it%20successfully.%20If%20this%20second%20read%20fails%20or%20the%20entry%20is%20absent%20%28narrow%20but%20possible%20TOCTOU%20between%20classify%20and%20here%29%2C%20%60actual%60%20silently%20falls%20back%20to%20%60source_version%60%2C%20producing%20%60manifest_expected_version_mismatch%28table_key%2C%20source_version%2C%20source_version%29%60%20%E2%80%94%20i.e.%20%60expected%20%3D%3D%20actual%60.%20The%20error%20is%20retryable%20and%20correct%2C%20but%20the%20diagnostic%20is%20confusing%3A%20a%20caller%20inspecting%20version%20numbers%20sees%20no%20mismatch.%20Using%20%60manifest_conflict%60%20%28like%20the%20%60Indeterminate%60%20arm%29%20avoids%20the%20misleading%20%60expected%20%3D%3D%20actual%60%20case%20when%20the%20version%20can't%20be%20confirmed.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20ForkRefStatus%3A%3ALegitimate%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20A%20concurrent%20writer%20published%20a%20real%20fork%20for%20this%20table%20on%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20active_branch.%20Retrieve%20the%20version%20from%20the%20snapshot%20we%20already%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20read%20in%20classify_fork_ref.%20If%20the%20second%20read%20fails%20%28narrow%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20TOCTOU%29%2C%20fall%20back%20to%20a%20manifest_conflict%20so%20the%20error%20never%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20claims%20expected%20%3D%3D%20actual%20and%20misleads%20callers.%0A%20%20%20%20%20%20%20%20%20%20%20%20let%20actual_version%20%3D%20db%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20.fresh_snapshot_for_branch%28Some%28active_branch%29%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20.await%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20.ok%28%29%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20.and_then%28%7Cs%7C%20s.entry%28table_key%29.map%28%7Ce%7C%20e.table_version%29%29%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20Err%28match%20actual_version%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20Some%28actual%29%20%3D%3E%20OmniError%3A%3Amanifest_expected_version_mismatch%28%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20table_key%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20source_version%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20actual%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20None%20%3D%3E%20OmniError%3A%3Amanifest_conflict%28format!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22branch%20'%7Bactive_branch%7D'%20already%20has%20a%20legitimate%20fork%20for%20table%20%5C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20'%7Btable_key%7D'%3B%20refresh%20and%20retry%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%29%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%29%3B%0A%20%20%20%20%20%20%20%20%7D%0A%60%60%60%0A%0A&repo=modernrelay%2Fomnigraph&pr=231&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a>

<sub>Reviews (13): Last reviewed commit: ["Merge remote-tracking branch &#39;origin/mai..."](https://github.com/modernrelay/omnigraph/commit/1eea371e07ed0c1868a85eb68e2e3a29f56fbbb1) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37501333)</sub>

### CODE REVIEW

File: crates/omnigraph/src/exec/mutation.rs [hunk 1]
```diff
@@ -741,14 +741,45 @@ impl Omnigraph {
...
+            });
+            if needs_fork {
+                let guards = self.write_queue().acquire_many(&touched).await;
+                Some((touched, guards))
+            } else {
+                None
+            }
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/exec/mutation.rs:779]: Fixed in d2bccd9. `touched_table_keys` now includes delete-cascade edges (derived from the catalog via from_type/to_type, mirroring `execute_delete_node`), so the up-front fork-queue acquisition covers them. Added regression test `branch_cascade_delete_forks_node_and_edges_under_held_queues`, and promoted `commit_all`'s coverage check from `debug_assert` to an all-builds typed error. Real gap our tests missed — we only drove the path via insert/load, never delete. Thanks.

File: crates/omnigraph/src/db/omnigraph.rs [hunk 2]
```diff
@@ -1487,15 +1495,29 @@ impl Omnigraph {
...
             source_branch,
             source_version,
             active_branch,
         )
-        .await
+        .await?
+        {
+            crate::storage_layer::ForkOutcome::Created(ds) => Ok(ds),
+            crate::storage_layer::ForkOutcome::RefAlreadyExists => {
+                table_ops::reclaim_orphaned_fork_and_refork(
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph.rs:1515]: Fixed in fedda78. `reclaim_orphaned_fork_and_refork` now re-derives its precondition from a FRESH manifest read (`fresh_snapshot_for_branch`, which bypasses the coordinator cache) immediately before force-deleting, and refuses with a retryable conflict if the table is legitimately on the branch. Correct regardless of caller snapshot staleness (including the branch-merge target swap). Good catch.

File: crates/omnigraph/src/exec/staging.rs [hunk 3]
```diff
@@ -512,7 +525,23 @@ impl StagedMutation {
...
+        let guards = match held_guards {
+            Some((acquired_keys, guards)) => {
+                debug_assert!(
+                    {
+                        let held: std::collections::HashSet<&(String, Option<String>)> =
+                            acquired_keys.iter().collect();
+                        queue_keys.iter().all(|k| held.contains(k))
+                    },
+                    "commit_all: held_guards must cover every touched-table queue key"
+                );
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/exec/staging.rs]: Fixed in d2bccd9. The held-guard coverage check is now an all-builds check that returns a typed `manifest_internal` error (fails the write safely) instead of a `debug_assert`, so a future divergence between the pre-computed touched set and the staged/inline set fails loudly rather than silently doing an unserialized write.

File: crates/omnigraph/src/db/omnigraph/table_ops.rs [hunk 4]
```diff
@@ -540,6 +545,71 @@ pub(super) async fn fork_dataset_from_entry_state(
...
+    db.storage()
+        .force_delete_branch(full_path, active_branch)
+        .await
+        .map_err(|e| {
+            // Lance refuses to delete a branch with dependent child branches
+            // even under force (`refs.rs` RefConflict "referenced by N
+            // versions"). Unreachable for a leaf first-write fork, but surface
+            // it actionably instead of looping.
+            if e.to_string().contains("referenced by") {
+                OmniError::manifest_conflict(format!(
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/table_ops.rs]: Addressed in fedda78. Loosened the match so it no longer depends on Lance's exact RefConflict prose; a typed `RefConflict` variant threaded through `force_delete_branch` is noted as the durable follow-up. (Path is near-unreachable for a leaf first-write fork.)

File: crates/omnigraph/src/db/omnigraph/optimize.rs [hunk 5]
```diff
@@ -634,7 +647,52 @@ pub async fn reconcile_orphaned_branches(db: &Omnigraph) -> Result<BranchReconci
...
+            };
+            if is_orphan {
+                orphans.push(branch);
+            }
+        }
+        // Children before parents (longest name first) so Lance's referenced-
+        // parent RefConflict cannot block reclamation.
+        orphans.sort_by(|a, b| b.len().cmp(&a.len()).then_with(|| a.cmp(b)));
+
+        for branch in orphans {
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/optimize.rs:772]: Fixed in 509a9c2. The reconciler now obeys the same invariant as the write-path reclaim: before `force_delete_branch` it acquires the per-(table, branch) write queue and re-validates under it from a FRESH read. An in-process first-write fork in its fork→publish window holds that queue, so cleanup now waits for it; once the writer publishes, the fresh re-check sees the table on the branch and skips it. One key held at a time (no lock-order inversion vs multi-table acquire_many writers). Cross-process writers remain the documented one-winner-CAS gap. Good catch — this race didn't exist in the old branch-name-based reconciler (a deleted branch can't have a live first-write), so the per-table broadening introduced it.

File: crates/omnigraph/src/table_store.rs [hunk 6]
```diff
@@ -299,26 +300,24 @@ impl TableStore {
...
+            // fully-formed manifest-unreferenced fork, or a phase-1-only Lance
+            // "zombie" — `create_branch` fails for both). The caller
+            // (`open_owned_dataset_for_branch_write`) has already re-read the
+            // live manifest under the held write queue and confirmed it does
+            // NOT place this table on `target_branch`, so the ref is reclaimable
+            // derived state, not authoritative. Return the typed signal and let
+            // the db layer reclaim + re-fork rather than guessing or hard-failing
+            // here. (A live committed fork is routed to a retryable conflict
+            // upstream, before we ever reach this fork path.)
+            return Ok(ForkOutcome::RefAlreadyExists);
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/table_store.rs]: Fixed in 3300109. `fork_branch_from_state` no longer maps every `create_branch` failure to `RefAlreadyExists`. On failure it now checks whether the ref actually exists (`list_branches`): only a genuinely pre-existing ref routes to reclaim; any other failure (I/O, version, Lance-internal) propagates with fidelity. It deliberately does NOT force-delete on a not-found-ref failure — that's indistinguishable from a transient error on a fresh create, and force-deleting there is the overreach the fresh-authority guard already removed. (A rarer phase-1-only Lance zombie surfaces as the propagated error for manual reclaim.)

File: crates/omnigraph/src/db/omnigraph/optimize.rs [hunk 7]
```diff
@@ -658,7 +671,77 @@ pub async fn reconcile_orphaned_branches(db: &Omnigraph) -> Result<BranchReconci
...
+                .acquire(&(table_key.clone(), Some(branch.clone())))
+                .await;
+            let still_orphan = match db.fresh_snapshot_for_branch(Some(&branch)).await {
+                Ok(snap) => snap
+                    .entry(&table_key)
+                    .map(|e| e.table_branch.as_deref() != Some(branch.as_str()))
+                    .unwrap_or(true),
+                // Branch absent from the manifest entirely (origin 1) → orphan.
+                Err(_) => true,
+            };
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/optimize.rs]: Fixed in 0f6da7b. The reconcile re-check no longer treats a read error as "still orphan": it goes through the shared `classify_fork_ref`, which returns `Indeterminate` on a transient read of a live branch (cleanup then skips and records it, never force-deletes) and only `Orphan` when fresh authority confirms the manifest does not reference the fork. Same don't-destroy-on-ambiguity rule as the write-path reclaim.

[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/optimize.rs]: Fixed in 0f6da7b. The reconcile re-check no longer treats a read error as "still orphan": it goes through the shared `classify_fork_ref`, which returns `Indeterminate` on a transient read of a live branch (cleanup then skips and records it, never force-deletes) and only `Orphan` when fresh authority confirms the manifest does not reference the fork. Same don't-destroy-on-ambiguity rule as the write-path reclaim.

File: crates/omnigraph/src/db/omnigraph/optimize.rs [hunk 8]
```diff
@@ -658,7 +671,103 @@ pub async fn reconcile_orphaned_branches(db: &Omnigraph) -> Result<BranchReconci
...
+            let _guard = db
+                .write_queue()
+                .acquire(&(table_key.clone(), Some(branch.clone())))
+                .await;
+            let still_orphan = if !live_branches.contains(&branch) {
+                // Origin 1: the branch is absent from the manifest authority
+                // entirely — a confirmed orphan. No live writer can hold this
+                // branch's queue (you cannot first-write to a branch the
+                // manifest does not have), so no fresh re-check is needed.
+                true
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/optimize.rs]: Fixed in 0f6da7b. The stale `live_branches` shortcut for origin-1 is gone — every candidate is now decided under the queue from a FRESH per-branch read via `classify_fork_ref`. A branch created after the sweep started resolves to its published fork → `Legitimate` → skipped (no destroy); a genuinely-absent branch → `Orphan` → reclaimed (ghost reclaim preserved); a transient read → `Indeterminate` → skipped. `live_branches` remains only as a cheap candidate pre-filter where over-inclusion is harmless. Both reclaim sites now share this one classifier so they can't drift again.

[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/optimize.rs]: Fixed in 0f6da7b. The stale `live_branches` shortcut for origin-1 is gone — every candidate is now decided under the queue from a FRESH per-branch read via `classify_fork_ref`. A branch created after the sweep started resolves to its published fork → `Legitimate` → skipped (no destroy); a genuinely-absent branch → `Orphan` → reclaimed (ghost reclaim preserved); a transient read → `Indeterminate` → skipped. `live_branches` remains only as a cheap candidate pre-filter where over-inclusion is harmless. Both reclaim sites now share this one classifier so they can't drift again.

File: crates/omnigraph/src/db/omnigraph/optimize.rs [hunk 9]
```diff
@@ -658,7 +671,94 @@ pub async fn reconcile_orphaned_branches(db: &Omnigraph) -> Result<BranchReconci
...
+                            tracing::warn!(
+                                target: "omnigraph::cleanup",
+                                table = %table_key,
+                                branch = %branch,
+                                error = %err,
+                                "resolving branch snapshot failed during reconcile; skipping",
+                            );
+                            stats.failures.push((table_key.clone(), err.to_string()));
+                            continue;
+                        }
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/optimize.rs:756]: Fixed in a9f39f5. The reconciler now caches failed live-branch snapshot resolutions in `failed_branch_snapshots`, so one branch-level read failure is recorded once and later tables that list the same branch skip without refetching or duplicating `stats.failures`. Added failpoint regression `reconcile_caches_live_branch_snapshot_resolution_failure` to pin the behavior.

File: crates/omnigraph/src/db/omnigraph/table_ops.rs [hunk 10]
```diff
@@ -623,6 +628,162 @@ pub(super) async fn fork_dataset_from_entry_state(
...
+            .fresh_snapshot_for_branch(Some(active_branch))
+            .await
+            .ok()
+            .and_then(|s| s.entry(table_key).map(|e| e.table_version))
+            .unwrap_or(source_version);
+        return Err(OmniError::manifest_expected_version_mismatch(
+            table_key,
+            source_version,
+            actual,
+        ));
```
[the remark this answers is not in the document]
Comment (ragnorc) [reply crates/omnigraph/src/db/omnigraph/table_ops.rs]: Fixed in 9e9cdbc. `reclaim_orphaned_fork_and_refork` now treats `ForkRefStatus::Indeterminate` as its own retryable manifest conflict that names the unavailable fresh authority read, rather than re-reading and reporting a misleading `ExpectedVersionMismatch`. Added failpoint regression `recreate_over_orphaned_fork_reports_indeterminate_authority_read` to pin that the fork is left untouched on ambiguity and converges once authority is readable.
